### ISSUE TITLE
We might not need Lodash

### ISSUE THREAD

**Issue by StuAA78** [DEFRA/water-abstraction-team#25]
TL;DR with some work we may be able to drop Lodash as a dependency:
* https://youmightnotneed.com/lodash
* https://you-dont-need.github.io/You-Dont-Need-Lodash-Underscore

---

We use [Lodash](https://www.npmjs.com/package/lodash) a _lot_ in our codebase! And this sometimes makes the purpose of code unclear. For example, in the UI the default mapper for converting form input into the internal datatype and back again looks like this:

```js
const defaultMapper = {
  import: (fieldName, payload) => {
    return payload[fieldName];
  },
  export: identity,
  postValidate: identity
};
```

Lodash's `identity` function returns the first value passed to it. We spent a little time wondering exactly why this was needed here in the `export` and `postValidate` hooks of the mapper! Eventually we realised that the purpose of the code is simply that when the `export` and `postValidate` hooks are called, and the value of the field is implicitly passed in, it simply returns that same value back. This is in contrast to the licence number mapper, for example:

```js
const licenceNumbersMapper = {
  import: (fieldName, payload) => {
    return extractLicenceNumbers(payload[fieldName]);
  },
  export: (value) => {
    return value.join(', ');
  },
  postValidate: identity
};
```

Here it's very clear that the `export` hook returns a joined list of licence numbers! So why didn't we just write the original mapper like this, which makes it very clear at a glance that the `export` and `postValidate` hooks return the values they're given:

```js
const defaultMapper = {
  import: (fieldName, payload) => {
    return payload[fieldName];
  },
  export: (value) => {
    return value;
  },
  postValidate: value => value
};
```

(`export` and `postValidate` do the same thing here, the latter was simply written as a one-liner to show how neat the code could be)

And this is simply one case where Lodash is being used where it needn't be!

There are resources which show how some Lodash functions can be replaced with vanilla JS:
* https://youmightnotneed.com/lodash
* https://you-dont-need.github.io/You-Dont-Need-Lodash-Underscore

So we may be able to drop Lodash as a dependency entirely, or at the very least rewrite some parts of the code to use vanilla JS where this helps us make the code clearer.

**Comment by StuAA78** [comment]
I only noticed after I wrote all this that we've already got something open for [deprecated packages](https://github.com/DEFRA/water-abstraction-team/issues/22) which includes Lodash, but thought I'd go ahead and open the issue anyway due to the extra detail -- it keeps in sight the fact that Lodash is hiding the function of some of the code so even if we end up keeping it (eg. because a Lodash method is genuinely useful and not just handy shorthand for vanilla JS) we should at least consider rewriting things where Lodash isn't helping.

**Comment by Cruikshanks** [comment]
💯 agree with your comment about opening a separate ticket for this and ...

![yes](https://user-images.githubusercontent.com/1789650/168772681-d022229b-e1a8-4483-8991-ce1d66e5d034.gif)

**Comment by Beckyrose200** [comment]
- [ ] - water-abstraction-admin-ui
- [ ] -  water-abstraction-import
- [ ] -  water-abstraction-permit-repository
- [ ] -  water-abstraction-reporting
- [ ] -  water-abstraction-returns
- [ ] -  water-abstraction-service
- [ ] -  water-abstraction-tactical-crm
- [ ] -  water-abstraction-tactical-idm
- [ ] -  water-abstraction-ui

**Comment by Cruikshanks** [comment]
@Beckyrose200 made a super-hero effort to remove lodash from all the legacy repos. They almost got there with just some true shockers left to tackle.

These final uses are as complicated as they are because lodash is doing too much and has been carelessly thrown about the place.

When the current dev team first joined Water Abstraction we started creating issues for all the problems we saw, this being a key one. Whilst we were still getting to grips with the service it gave us opportunities to fix things and get stuck into the legacy code.

After a while, we became more confident and were able to get stuck into the critical feature at that time; SROC Supplementary billing. It also proved we could implement new and existing functionality in our [water-abstraction-system repo](https://github.com/DEFRA/water-abstraction-system) and integrate it with the existing apps. This, and the ongoing performance and stability issues with the legacy code have meant the whole delivery team are now onboard with us, over time, migrating _all_ functionality across.

So, as a dev team, we're instead going to focus our efforts on migrating functionality rather than fixing the legacy code. It means issues like this are now redundant as we don't intend to ever fix them.

Still, well done @Beckyrose200 ! 💪🏅
